proof/refine/Invariants_H.thy

```diff
@@ -2148,109 +1778,23 @@ lemma tcb_ctes_clear:
-   apply (simp add: loadObject_cte unless_def alignCheck_def
-                    is_aligned_mask[symmetric] objBits_simps'
-                    cte_level_bits_def magnitudeCheck_def
-                    return_def fail_def)
-   apply (clarsimp simp: bind_def return_def when_def fail_def
-                  split: option.splits)
-   apply simp
-  apply (erule(1) ps_clear_lookupAround2)
-    prefer 3
-    apply (simp add: loadObject_cte unless_def alignCheck_def
-                     is_aligned_mask[symmetric] objBits_simps'
-                     cte_level_bits_def magnitudeCheck_def
-                     return_def fail_def tcbCTableSlot_def tcbVTableSlot_def
-                     tcbIPCBufferSlot_def tcbReplySlot_def tcbCallerSlot_def
-              split: option.split_asm)
-     apply (clarsimp simp: bind_def tcb_cte_cases_def cteSizeBits_def split: if_split_asm)
-    apply (clarsimp simp: bind_def tcb_cte_cases_def iffD2[OF linorder_not_less]
-                          return_def cteSizeBits_def
-                   split: if_split_asm)
-   apply (subgoal_tac "p - n \<le> (p - n) + n", simp)
-   apply (erule is_aligned_no_wrap')
-    apply (simp add: word_bits_conv)
-   apply (simp add: tcb_cte_cases_def cteSizeBits_def split: if_split_asm)
-  apply (subgoal_tac "(p - n) + n \<le> (p - n) + 0x7FF")
-   apply (simp add: field_simps)
-  apply (rule word_plus_mono_right)
-   apply (simp add: tcb_cte_cases_def cteSizeBits_def split: if_split_asm)
-  apply (erule is_aligned_no_wrap')
-  apply simp
-  done
 
-lemma tcb_at_cte_at':
-  "tcb_at' t s \<Longrightarrow> cte_at' t s"
-  apply (clarsimp simp add: cte_wp_at_cases' obj_at'_def projectKO_def
-                       del: disjCI)
-  apply (case_tac ko)
-   apply (simp_all add: projectKO_opt_tcb fail_def)
-  apply (rule exI[where x=0])
-  apply (clarsimp simp add: return_def objBits_simps)
-  done
-
-lemma cte_wp_atE' [consumes 1, case_names CTE TCB]:
-  assumes cte: "cte_wp_at' P ptr s"
-  and   r1: "\<And>cte.
-    \<lbrakk> ksPSpace s ptr = Some (KOCTE cte); ps_clear ptr cte_level_bits s;
-      is_aligned ptr cte_level_bits; P cte \<rbrakk> \<Longrightarrow> R"
-  and   r2: "\<And> tcb ptr' getF setF.
-  \<lbrakk> ksPSpace s ptr' = Some (KOTCB tcb); ps_clear ptr' tcbBlockSizeBits s; is_aligned ptr' tcbBlockSizeBits;
-     tcb_cte_cases (ptr - ptr') = Some (getF, setF); P (getF tcb) \<rbrakk> \<Longrightarrow> R"
-  shows "R"
-  by (rule disjE [OF iffD1 [OF cte_wp_at_cases' cte]]) (auto intro: r1 r2 simp: cte_level_bits_def objBits_defs)
-
-lemma cte_wp_at_cteI':
-  assumes "ksPSpace s ptr = Some (KOCTE cte)"
-  assumes "is_aligned ptr cte_level_bits"
-  assumes "ps_clear ptr cte_level_bits s"
-  assumes "P cte"
-  shows "cte_wp_at' P ptr s"
-  using assms by (simp add: cte_wp_at_cases' cte_level_bits_def objBits_defs)
-
-lemma cte_wp_at_tcbI':
+(* FIXME arch-split: naming: do we want to use the opaque naming style of Invariants_AI_2? *)
+locale Invariants_H_cte_ats =
```

Comment:
I like proper names better, at least if they make sense. Sometimes there is too much stuff in there for a good name, but `cte_ats` here is good.

Reply:
In this case, if we can make `tcb_cte_cases` generic, I suspect that this locale becomes unneccessary.

Reply:
I don't know, `cte_wp_at_cases'` is pretty nasty and I don't have the time available to go through it carefully... not seeing a sensible solution inside.

Reply:
No worries, fine to leave for another time when we have better infrastructure for it.